core/deliverservice/bft_client.go

```diff
@@ -0,0 +1,546 @@
+		minBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.minBackoffDelay", bftMinBackoffDelay),
+		maxBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.maxBackoffDelay", bftMaxBackoffDelay),
+		blockCensorshipTimeout:              util.GetDurationOrDefault("peer.deliveryclient.bft.blockCensorshipTimeout", bftBlockCensorshipTimeout),
+		endpoints:                           comm.Shuffle(endpoints),
+		blockReceiverIndex:                  -1,
+		headerReceivers:                     make(map[string]*bftHeaderReceiver),
+	}
+
+	bftLogger.Infof("[%s] Created BFT Delivery Client", chainID)
+	return c
+}
+
+func (c *bftDeliveryClient) Recv() (response *orderer.DeliverResponse, err error) {
+	bftLogger.Debugf("[%s] Entry", c.chainID)
+
+	c.startOnce.Do(func() {
+		var num uint64
+		num, err = c.ledgerInfoProvider.LedgerHeight()
+		if err != nil {
+			return
+		}
+		c.nextBlockNumber = num
+		c.lastBlockTime = time.Now()
+		bftLogger.Debugf("[%s] Starting monitor routine; Initial ledger height: %d", c.chainID, num)
+		go c.monitor()
+	})
+	// can only happen once, after first invocation
+	if err != nil {
+		bftLogger.Debugf("[%s] Exit: error=%v", c.chainID, err)
+		return nil, errors.New("cannot access ledger height")
+	}
+
+	var numEP int
+	var numRetries int
+	var stopRetries = time.Now().Add(c.reconnectTotalTimeThreshold)
+
+	for !c.shouldStop() {
+		if numEP, err = c.assignReceivers(); err != nil {
+			bftLogger.Debugf("[%s] Exit: error=%s", c.chainID, err)
+			return nil, err
+		}
+
+		c.launchHeaderReceivers()
+
+		response, err = c.receiveBlock()
+		if err == nil {
+			bftLogger.Debugf("[%s] Exit: response=%v", c.chainID, response)
+			return response, nil // the normal return path
+		}
+
+		if stopRetries.Before(time.Now()) {
+			bftLogger.Debugf("[%s] Exit: reconnectTotalTimeThreshold: %s, expired; error: %s",
+				c.chainID, c.reconnectTotalTimeThreshold, errClientReconnectTimeout)
+			return nil, errClientReconnectTimeout
+		}
+
+		c.closeBlockReceiver()
+		numRetries++
+		if numRetries%numEP == 0 { //double the back-off delay on every round of attempts.
+			dur := backOffDuration(2.0, uint(numRetries/numEP), c.minBackoffDelay, c.maxBackoffDelay)
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint in: %s", c.chainID, err, dur)
+			backOffSleep(dur, c.stopChan)
+		} else {
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint now", c.chainID, err)
+		}
+	}
+
+	bftLogger.Debugf("[%s] Exit: %s", c.chainID, errClientClosing.Error())
+	return nil, errClientClosing
+}
+
+func backOffDuration(base float64, exponent uint, minDur, maxDur time.Duration) time.Duration {
+	if base < 1.0 {
+		base = 1.0
+	}
+	if minDur <= 0 {
+		minDur = bftMinBackoffDelay
+	}
+	if maxDur < minDur {
+		maxDur = minDur
+	}
+
+	fDurNano := math.Pow(base, float64(exponent)) * float64(minDur.Nanoseconds())
```

Comment:
Not sure how safe this is. 
We should IMO multiply by the exponent by the amount of milliseconds in minDuration, and then after the exponentiation we can just convert it to absolute time.Duration.

Reply:
This fine. even if after exponentiation fDurNano is +Inf, the next line 

fDurNano = math.Min(fDurNano, float64(maxDur.Nanoseconds()))

will convert back to the maxDur, wich is non Inf.
this is because in go: math.Min(val, +Inf) = val